equinix/helpers_device.go

```diff
@@ -271,3 +319,42 @@ func ipAddressSchema() *schema.Resource {
 		},
 	}
 }
+
+func getDeviceMap(device metalv1.Device) map[string]interface{} {
+	networkInfo := getNetworkInfoMetalGo(device.IpAddresses)
+	sort.SliceStable(networkInfo.Networks, func(i, j int) bool {
+		famI := int(*networkInfo.Networks[i]["family"].(*int32))
+		famJ := int(*networkInfo.Networks[j]["family"].(*int32))
+		pubI := *networkInfo.Networks[i]["public"].(*bool)
+		pubJ := *networkInfo.Networks[j]["public"].(*bool)
+		return getNetworkRank(famI, pubI) < getNetworkRank(famJ, pubJ)
+	})
+	keyIDs := []string{}
+	for _, k := range device.SshKeys {
+		keyIDs = append(keyIDs, path.Base(k.GetHref()))
+	}
+	ports := getPortsMetalGo(device.NetworkPorts)
+
+	return map[string]interface{}{
+		"hostname":            device.GetHostname(),
+		"project_id":          *device.GetProject().Id,
```

Comment:
Is this possible, for this line and the others, as a way to have safer dereferencing:

```suggestion
		"project_id":          device.GetProject().GetId(),
```

Reply:
Chained lookups should look like this:
```suggestion
		"project_id":          device.Project.GetId(),
```

`device.Project` is a `*Project`, which has [`GetId()` with the necessary nil checks](https://github.com/equinix-labs/metal-go/blob/main/metal/v1/model_project.go#L226-L232)

`device.GetProject()` returns a `Project`, which doesn't have `GetId()`

Reply:
the `Project` value can use the `*Pointer` pointer receivers: https://go.dev/tour/methods/4

Reply:
I haven't tried it specifically in this PR, but when I tried [the similar suggestion on a different PR](https://github.com/equinix/terraform-provider-equinix/pull/369#discussion_r1320210247) I got a build error `equinix/helpers_device.go:176:26: cannot call pointer method GetMac on v1.PortData`.  What am I doing wrong?

Reply:
It looks like I'm not doing anything wrong, and the error is expected in this scenario: https://stackoverflow.com/a/44543748

Summarizing from there, the return value of `GetProject()` isn't addressable, so Go can't automatically pointerize it to look up the pointer receiver as a convenience.

Since the `GetX()` functions are written to handle a nil receiver, and since this function receives a `Device` and not a `*Device`, the `device.Project.GetId()` call here is nil-safe.  IMO there's no benefit to doing the extra work we would need to do in order to get a pointer for the value returned by `device.GetProject()`.

Reply:
Interesting and well enough. I tried a few combinations myself: https://go.dev/play/p/YcDgg2oELju (go.dev won't download metal-go successfully)